src/rules.jl

```diff
@@ -1,260 +1,313 @@
-                 end
-             else
-                 msg = "The phase of spider $(v) is not 0!"
-             end
-        else
-            msg = "Spider $(v) is not a Z or X spider!"
+function match_rule(::Rule_pi, zxd::ZXDiagram{T, P}) where {T, P}
+    matches = Match{T}[]
+    for v1 in spiders(zxd)
+        if spider_type(zxd, v1) == X && phase(zxd, v1) == one(P) && length(neighbors(zxd, v1, count_mul = true)) == 2
+            for v2 in neighbors(zxd, v1)
+                if spider_type(zxd, v2) == Z
+                    push!(matches, Match{T}([v1, v2], Rule_pi()))
+                end
+            end
         end
-    else
-        msg = "Spider $(v) is not in this ZX-diagram!"
     end
-    return msg
+    return matches
 end
 
-function rule_i1!(zxd::ZXDiagram{T,U,P}, v::T) where {T<:Integer, U<:Integer, P}
-    msg = check_rule_i1(zxd, v)
-    if msg != true
-        error(msg)
+function match_rule(::Rule_c, zxd::ZXDiagram{T, P}) where {T, P}
+    matches = Match{T}[]
+    for v1 in spiders(zxd)
+        if spider_type(zxd, v1) == X && phase(zxd, v1) == zero(P) && length(neighbors(zxd, v1, count_mul = true)) == 1
+            for v2 in neighbors(zxd, v1)
+                if spider_type(zxd, v2) == Z
+                    push!(matches, Match{T}([v1, v2], Rule_c()))
+                end
+            end
+        end
     end
-    nbhd = outneighbors(zxd, v)
-    if length(nbhd) == 2
-        add_edge!(zxd.g, nbhd[1], nbhd[2])
+    return matches
+end
+
+function match_rule(::Rule_b, zxd::ZXDiagram{T, P}) where {T, P}
+    matches = Match{T}[]
+    for v1 in spiders(zxd)
+        if spider_type(zxd, v1) == X && phase(zxd, v1) == zero(P) && length(neighbors(zxd, v1, count_mul = true)) == 3
+            for v2 in neighbors(zxd, v1)
+                if spider_type(zxd, v2) == Z && phase(zxd, v2) == zero(P) && length(neighbors(zxd, v2, count_mul = true)) == 3 && mul(zxd.mg, v1, v2) == 1
+                    push!(matches, Match{T}([v1, v2], Rule_b()))
+                end
+            end
+        end
     end
-    vmap = rem_spider!(zxd, v)
+    return matches
+end
 
-    return vmap
+function rewrite!(zxd::ZXDiagram{T, P}, matches::Vector{Match{T}}) where {T, P}
+    for each in matches
+        rewrite!(zxd, each)
+    end
+    zxd
 end
 
-function check_rule_i2(zxd::ZXDiagram{T,U,P}, v1::T, v2::T) where {T<:Integer, U<:Integer, P}
-    msg = true
-    if v1 <= nv(zxd) && v2 <= nv(zxd)
-        if !(zxd.st[v1] == H && degree(zxd.g, v1) == 2)
-            msg = "Spider $(v1) is not an H spider!"
-        end
-        if !(zxd.st[v2] == H && degree(zxd.g, v2) == 2)
-            msg = "Spider $(v2) is not an H spider!"
-        end
-        if !has_edge(zxd.g, v1, v2)
-            msg = "Spiders $(v1) and $(v2) are not connected!"
-        end
+function rewrite!(zxd::ZXDiagram{T, P}, matched::Match{T}) where {T, P}
+    r = matched.rule
+    vs = matched.vertices
+    if check_rule(zxd, r, vs)
```

Comment:
do we actually need to check this in `rewrite!`? remember the structure in `simplify`

```julia
m = match(r, diagram) # m::Match
if m !== nothing
    m = filter_overlapped_graph(m)
    rewrite!(r, diagram, m)
end
```

so we always know the rule matches, there is no need to check the rules here.